src/smt.rs

```diff
@@ -1,163 +1,625 @@
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum SMTSort {
+    Bool,
+    BV(u64),
+    Array(Box<SMTSort>, Box<SMTSort>),
+}
+
+pub fn bv(width: u64) -> SMTSort {
+    SMTSort::BV(width)
+}
+
+pub fn array(index: SMTSort, elem: SMTSort) -> SMTSort {
+    SMTSort::Array(Box::new(index), Box::new(elem))
+}
+
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub struct SMTVariable {
     pub name: String,
+    pub sort: SMTSort,
 }
 
-impl SMTFormat for SMTVariable {
-    fn as_smt(&self) -> String {
-        self.name.clone()
+impl SMTVariable {
+    pub fn new(name: String, sort: SMTSort) -> Self {
+        SMTVariable { name, sort }
     }
 }
 
-pub trait SMTFormat {
-    fn as_smt(&self) -> String;
+/// We keep the SMT expressions loose and runtime based for now.
+/// A potential future refactor may add compile time guarantees
+/// that illegal expressions cannot be built.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct SMTExpr {
+    pub op: SMTOp,
+    pub args: Vec<SMTExpr>,
 }
 
-type BoxedSMTFormat = Box<dyn SMTFormat>;
+impl From<u64> for SMTExpr {
+    fn from(input: u64) -> SMTExpr {
+        SMTExpr {
+            op: SMTOp::Literal(format!("{:064x}", input)),
+            args: vec![],
+        }
+    }
+}
 
-impl SMTFormat for String {
-    fn as_smt(&self) -> String {
-        self.to_string()
+impl From<SMTVariable> for SMTExpr {
+    fn from(input: SMTVariable) -> SMTExpr {
+        SMTExpr {
+            op: SMTOp::Variable(input),
+            args: vec![],
+        }
     }
 }
 
-impl SMTFormat for usize {
-    fn as_smt(&self) -> String {
-        format!("#x{:064x}", self)
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum SMTStatement {
+    Assert(SMTExpr),
+    DeclareConst(SMTVariable),
+    DeclareFun(SMTVariable, Vec<SMTSort>),
+    DefineConst(SMTVariable, SMTExpr),
+    DefineFun(SMTVariable, Vec<SMTSort>, SMTExpr),
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum SMTOp {
+    Eq,
+    Not,
+    And,
+    Or,
+    Ite,
+    BvNot,
+    BvAnd,
+    BvOr,
+    BvXor,
+    BvAdd,
+    BvSub,
+    BvMul,
+    BvUDiv,
+    BvSDiv,
+    BvURem,
+    BvSMod,
+    BvULt,
+    BvUGt,
+    BvSLt,
+    BvSGt,
+    BvShl,
+    BvLShr,
+    BvAShr,
+    Extract(u64, u64),
+    Concat,
+    Select,
+    Store,
+    Literal(String),
+    Variable(SMTVariable),
+    UF(SMTVariable), // TODO We should have a specialized SMTFunction
+}
+
+// SMT expression builders
+
+pub fn eq<L: Into<SMTExpr>, R: Into<SMTExpr>>(lhs: L, rhs: R) -> SMTExpr {
+    SMTExpr {
+        op: SMTOp::Eq,
+        args: vec![lhs.into(), rhs.into()],
     }
 }
 
-impl<T: SMTFormat> SMTFormat for Vec<T> {
-    fn as_smt(&self) -> String {
-        self.iter()
-            .map(SMTFormat::as_smt)
-            .collect::<Vec<_>>()
-            .join(" ")
+pub fn neq<L: Into<SMTExpr>, R: Into<SMTExpr>>(lhs: L, rhs: R) -> SMTExpr {
```

Comment:
Does the following also work? I think it is just syntactic sugar for what you currently have.
```suggestion
pub fn neq(lhs: impl Into<SMTExpr>, rhs: impl Into<SMTExpr>) -> SMTExpr {
```